keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,950 @@
+
+## User Stories 
+
+### Story 1: Feature Rollout Challenges with Version Skew
+
+To maintain cluster stability and enable safer rollouts, cluster administrators perform gradual upgrades. This inherently creates a mixed-version Kubernetes cluster, where nodes can be on different versions than the control plane. This environment introduces significant challenges for both pod scheduling and API request validation. The primary concerns in such clusters are:
+
+1. How do we prevent new pods that might use newer features from being scheduled on older, incompatible nodes?
+  * Example: [Pod Level Resources](https://github.com/kubernetes/enhancements/blob/63d4f6f2aa0e2eb0b83067b067c4949643b1b24c/keps/sig-node/2837-pod-level-resource-spec/README.md?plain=1#L4)
+    * The existing beta feature lacks explicit version-skew management, requiring it to be enabled on all components (scheduler, kubelet, API server) to function correctly. If enabled only on the control plane, a pod can be scheduled on an incompatible node where the Kubelet will reject it during admission. This rejection puts the pod into a `Failed` state, preventing it from being retried on other newer (compatible) nodes in the cluster.
+    * For new enhancements, like adding support for CPU alignment (static CPU policy support) there is no way to make sure a guaranteed QOS pod using pod level resources lands on a node supporting this new feature.
+
+2. How do we block API calls that target pods on nodes that don't support newer features?
+  * Example: [In-Place Pod Resizing](https://kubernetes.io/docs/tasks/configure-pod-container/resize-container-resources/)
+    * For an existing beta feature, there currently exists a [workaround](https://github.com/kubernetes/kubernetes/blob/23258f104d74c6f27fd4db94940d745d9d463a8f/pkg/apis/core/validation/validation.go#L5796) to handle this version skew by looking for alternate signals from the pod spec. However such workarounds may be complex and not always feasible for new features.
+    * For a new feature enhancement like [In-Place Pod Resize for Guaranteed QoS pods](https://github.com/kubernetes/enhancements/issues/5294), an API request to modify a running pod must be validated. The operation should be rejected if the pod resides on an older node that does not support the feature.
+
+The Node Capabilities framework addresses both issues through a unified mechanism without requiring any intervention (like adding Node Labels, selectors etc.). The scheduler uses the capability information to correctly filter nodes for new pods, while the admission controller validates operations against the capabilities of a pod's current node. This ensures that feature incompatibilities are handled proactively at the control plane.
+
+## Design Details
+
+### API Changes
+
+Add a `Capabilities` field as type `map[string]string` to the `Node.Status` structure. 
```

Comment:
will we allow editing the list of capabilities after node was registered? I assume yes since we are saying earlier that kubelet restart is needed to update capabilities, just wanted to confirm

Reply:
Yes. The node.status.capabilities field can be updated after the node is registered. The capabilitities are determined once during kubelet startup, and we need a kubelet restart for new capabilitied to be discovered and published.